pkg/scheduler/actions/preempt/preempt.go

```diff
@@ -128,8 +128,8 @@ func (alloc *Action) Execute(ssn *framework.Session) {
 				}
 			}
 
-			// Commit changes only if job is pipelined, otherwise try next job.
-			if ssn.JobPipelined(preemptorJob) {
+			// Commit changes only if plugin PreemptCommit fn allow, otherwise try next job.
+			if ssn.PreemptCommit(preemptorJob) {
```

Comment:
What does `preemptCommit` mean? Callback function should not bind to action.

Reply:
btw, how do we define `Pipelined` state for batch & elastic job?

Reply:
Please try something like this :)

```
@@ -272,14 +272,21 @@ func (bp *tdmPlugin) OnSessionOpen(ssn *framework.Session) {

        jobPipelinedFn := func(obj interface{}) bool {
                jobInfo := obj.(*api.JobInfo)
-               if jobInfo.Preemptable {
-                       // ignore preemptable job
-                       return true
-               }
                // non preemptable job(high priority) try to feed all its tasks
                occupied := jobInfo.WaitingTaskNum() + jobInfo.ReadyTaskNum()
-               return occupied >= (int32)(len(jobInfo.Tasks))
+               return occupied >= jobInfo.MinAvailable
+       }
+
+       jobStarvingFn := func(obj interface{}) bool {
+               jobInfo := obj.(*api.JobInfo)

+               // For preemptable job, it's nice to have more resource, but never starving.
+               if jobInfo.Preemptable {
+                       return false
+               }
+
+               occupied := jobInfo.WaitingTaskNum() + jobInfo.ReadyTaskNum()
+               return occupied < int32(len(jobInfo.Tasks))
        }

        ssn.AddPredicateFn(bp.Name(), predicateFn)

```

Reply:
Address comments: introduce `jobStarving` function, which check if job still need more resource. And `jobPipelined` function will check if the above preemtption can be commited to ssn.